### DISCUSSION TITLE
How are image props handled?

### DISCUSSION THREAD

**Discussion by grahammendick** [reactwg/react-native-new-architecture#31] [Q&A] [answered]
I've started updating my navigation library to the new architecture on iOS. I'm working on the `BarButton` component that maps to a `UIBarButtonItem`. It has an image prop,

```jsx
<BarButton image={require('./home.png')} />
```
In the old architecture on iOS I'd map this to a `UIImage` property and ReactNative would do the rest.

```obj-c
RCT_EXPORT_VIEW_PROPERTY(image, UIImage)

- (void)setImage:(UIImage *)image
{
    self.button.image = image;
}
```
What's the way to do this in the new architecture on iOS?

**Comment by cipolleschi** [comment]
Hi @grahammendick, thanks for your question.

In the New Architecture, we still need this snippet in the Native Component. 

```c++
RCT_EXPORT_VIEW_PROPERTY(image, UIImage)
```

Then, in the Fabric Component, we have to implement the following method:

```c++
- (void)updateProps:(Props::Shared const &)props oldProps:(Props::Shared const &)oldProps
```

An example on how to do that can be found [here](https://github.com/react-native-community/RNNewArchitectureLibraries/tree/feat/back-fabric-comp#fabric-component-update-the-native-ios-code) at step 5.

The likely implementation, but please test it, will be something like:

```c++
- (void)updateProps:(Props::Shared const &)props oldProps:(Props::Shared const &)oldProps
{
    const auto &oldViewProps = *std::static_pointer_cast<BarButtonProps const>(_props);
    const auto &newViewProps = *std::static_pointer_cast<BarButtonProps const>(props);

    if (oldViewProps.image != newViewProps.image) {
        _view.button.image = newViewProps.image; 
    }

    [super updateProps:props oldProps:oldProps];
}
```

Let me know if it works or if you had any other issue!

**Reply by grahammendick** [reply]
That works for props that are strings and booleans, for example. But I'm asking about a prop that is an image. So on the native side how to I turn this into a `UIImage`?
```jsx
<BarButton image={require('./home.png')} />
```
In the old architecture React Native would handle this automatically but I don't think it does anymore. I looked at the the `RCTSlider` and it is [using an ImageResponseCoordinatorObserver](https://github.com/facebook/react-native/blob/29c5461114988b16a72748be7caa0178f95e0908/React/Fabric/Mounting/ComponentViews/Slider/RCTSliderComponentView.mm#L162) for its `maximumTrackImage` prop. But I couldn't follow the code at all.

**Reply by cipolleschi** [reply]
## How the code is executed

- `getCoordinator` is a lambda that, given an `ImageRequest`, returns the corresponding coordniator
- the next ifs get the coordinators for the corresponding image if there is no data (first loading) or if the image changes.
- the previous state is loaded at the beginning of that method, and it is the SliderShadowNode::ConcreteState. You can see the generated code for the [slider here ](https://github.com/facebook/react-native/tree/8bd3edec88148d0ab1f225d2119435681fbbba33/ReactCommon/react/renderer/components/slider).
- The `ImageResponseCoordinators` can be found [here](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/ReactCommon/react/renderer/imagemanager/ImageResponseObserverCoordinator.cpp). As far as I can see, as soon as they get an observer, they start issueing a request.
- The various `setXXXImageCoordinator` method (like [this](https://github.com/facebook/react-native/blob/29c5461114988b16a72748be7caa0178f95e0908/React/Fabric/Mounting/ComponentViews/Slider/RCTSliderComponentView.mm#L198)) sets the coordinators and the observers, thus issueing the request.
- The observer is a Proxy object. The variable is declared [here](https://github.com/facebook/react-native/blob/29c5461114988b16a72748be7caa0178f95e0908/React/Fabric/Mounting/ComponentViews/Slider/RCTSliderComponentView.mm#L39), populated [here](https://github.com/facebook/react-native/blob/29c5461114988b16a72748be7caa0178f95e0908/React/Fabric/Mounting/ComponentViews/Slider/RCTSliderComponentView.mm#L60) - where it also passes the object itself as a delegate. The class is defined [here](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/Fabric/RCTImageResponseObserverProxy.mm).
- If I'm not mistaken, when the image is ready, the proxy object invokes [this method](https://github.com/facebook/react-native/blob/29c5461114988b16a72748be7caa0178f95e0908/React/Fabric/Mounting/ComponentViews/Slider/RCTSliderComponentView.mm#L326) which finally load the image.


This should be, more or less, the execution of what happens when we set an image. What's missing here is the type of the `require(./home.png)`. I think that from [here](https://github.com/facebook/react-native/blob/971ba5c26b335310944e0940c24ed9a01be7a365/Libraries/Components/Slider/Slider.js#L39) we can say that the image type is indeed an `ImageSource`.

## How to Implement the BarButton

If you want to mirror the slider approach, I think you have to:
1. declare your own `ImageResponseObserverCoordinator` and your own `RCTImageResponseObserverProxy`
2. make the `BarButton` conform to the `RCTImageResponseDelegate`.
3. Initialize the `proxy` in the `BarButton` init
4. in the `updateProps`, implement an `if` like the one you found in the slider.
5. Define your own `setImage` with the `removeObserver` - `coordinator assignment` - `addObserver` dance
6. Implement the `- (void)didReceiveImage:(UIImage *)image metadata:(id)metadata fromObserver:(void const *)observer` method.

I won't exclude that you can implement all these mechanisms in a different way, but I'm not sure how, at the moment.

---

I hope this helps, let me know if I can help you?

**Reply by grahammendick** [reply]
How do I go about generating `ShadowNode::ConcreteState` for my component? Is this documented anywhere? I can't find any documentation for `updateState` at all

**Reply by cipolleschi** [reply]
Yeah, we are working on that side as well.
As for now, the best resource is the [`Slider` spec](https://github.com/facebook/react-native/blob/971ba5c26b335310944e0940c24ed9a01be7a365/Libraries/Components/Slider/Slider.js).
You can also look at the [`SliderNativeComponent.js`](https://github.com/facebook/react-native/blob/971ba5c26b335310944e0940c24ed9a01be7a365/Libraries/Components/Slider/SliderNativeComponent.js).

In your specific case, I don't think you actually need to specify the State. The `BarButton`'s `image` property should be a `prop` for the component.

**Reply by grahammendick** [reply]
Why does the Slider need State for its image props if you say my component doesn't? What's different?

**Reply by cipolleschi** [reply]
I'm not saying you don't need it, but I don't think you don't have to specify it.
The Slider does not specify the state, neither in the Slider.js or in the SliderNativeComponent.js.

My guess is that as soon as you specify a prop with the type `ImageSource` something will generate the code you need. However, I can't try it right now... Could you create a simple spec for the `BarButton` with a prop of type ImageSource and see what happens?

**Reply by grahammendick** [reply]
I [tried that](https://github.com/grahammendick/navigation/blob/4aba99889d3769f17b32cad73077db0f23c36b86/NavigationReactNative/src/BarButtonNativeComponent.js#L15) already but it didn't work

**Reply by cipolleschi** [reply]
Ok, I think I'll need a bit of time to figure out what's happening and how that can be done.

**Comment by cipolleschi** [comment]
Hi @grahammendick! 

First of all, infinite thanks for your patience. It has been a busy period and I also had to reach out to different people to have a precise answer for your case. We can suggest 3 different ways to approach the solution.

## JS Based Approach
Let's start with saying that the ideal usage for React Native is to try to minimize the amount of native code, to share as much as JS as possible. Starting from this idea, the best and suggested approach to develop a component like the `<BarButton>` is actually to write that component in JS and to use an `<Image>` child. By following this approach, you will benefit from all the optimizations we develop at Meta and you will be able to share as much code as possible. This also has the best Developer Experience, benefitting from the fast reload and all the other JavaScript based features.

## Custom ImageLoader
The second approach by simplicity, is to create an `ImageLoader` in your native code and to use that to load the image. Roughly, the approach will be:

1. In the JS file, expose an `ImageSource` prop
2. In your `ViewManager.mm`, use the `RCT_EXPORT_VIEW_PROPERTY(<#name#>, <#type#>)` macro
3. Generate the code
4. Implement the `updateProps` function in the `View.mm` like this:
    ```c++
    - (void)updateProps:(Props::Shared const &)props oldProps:(Props::Shared const &)oldProps
    {
        const auto &oldViewProps = *std::static_pointer_cast<ColoredViewProps const>(_props);
        const auto &newViewProps = *std::static_pointer_cast<ColoredViewProps const>(props);
    
        if (oldViewProps.image != newViewProps.image) {
          const auto newURI = newViewProps.image.uri;
          [ImageLoader.shared loadImageFromUri:newUri];
        }
    
        [super updateProps:props oldProps:oldProps];
    }
    ```

**PRO:**
* Simpler path
* You own the whole code to load images
* You can leverage some external dependencies you may already depend upon
* You don't have to go through all the details of how ReactNative handle images

**CON:**
* You write your own code/depends on an additional dependency
* You won't leverage all the optimization Meta will develop for image loading

## Write Your Own State
Remember that the CodeGen is just a tool we created to simplify the 90% of the use cases. You don't have to use it if your use case is not well supported and, unfortunately, that is one of these cases. 

**What does that means?**
It means that you'll have to write manually a few files:

* The `BarButtonState.h` and the `BarButtonState.cpp`
* The `BarButtonProps.h` and the `BarButtonProps.cpp`
* The `BarButtonEventEmitter.h` and `BarButtonEventEmitter.cpp`
* The `BarButtonShadowNode.h` and the `BarButtonShadowNode.cpp`
* The `ComponentDescriptors.h`
* The `RCTComponentViewHelpers.h`

Once we have these elements, we can create our own `ColoredView` component.

We agree that this is not the best possible Dev X. We have plan to improve the Dev X and the codegen in the future, but we don't have a roadmap yet.

It follows the code of an example I created to verify that everything works.
For the example, I used a component I have ready, called `ColoredView`. I added the requirement of loading an image using the RN infrastructure.
It should be easy to follow the example, changing the bit you need to change.

**Note:** It's possible to partially leverage the codegen to get the scaffolding for many files.

---

### JS Spec
As always, the first step is to declare some JS specifications that we can leverage.

```typescript
// @flow
import type {ViewProps} from 'react-native/Libraries/Components/View/ViewPropTypes';
import type {HostComponent} from 'react-native';
import { ViewStyle } from 'react-native';
import type {ImageSource} from '../../Image/ImageSource';
import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';

type NativeProps = $ReadOnly<{|
    ...ViewProps,
    color: string,
    image?: ?ImageSource,
|}>;

export default (codegenNativeComponent<NativeProps>(
    'ColoredView',
): HostComponent<NativeProps>);
```

After that, we Have to create all the files that the codegen usually creates for us. We will have to update some of them, so the suggestion is to:
1. Generate them
2. Copy and paste them in your library
3. Modify them

So... Let's start:

### `ColoredViewState.h`
```c++
#pragma once

#ifdef ANDROID
#include <folly/dynamic.h>
#include <react/renderer/mapbuffer/MapBuffer.h>
#include <react/renderer/mapbuffer/MapBufferBuilder.h>
#endif

#include <react/renderer/imagemanager/ImageRequest.h>
#include <react/renderer/imagemanager/primitives.h>

namespace facebook {
namespace react {

/*
 * State for <Slider> component.
 */
class ColoredViewState final {
 public:
  ColoredViewState(
      ImageSource const &imageSource,
      ImageRequest imageRequest)
      : imageSource_(imageSource),
        imageRequest_(
            std::make_shared<ImageRequest>(std::move(imageRequest))){};

  ColoredViewState() = default;

  ImageSource getImageSource() const;
  ImageRequest const &getImageRequest() const;

#ifdef ANDROID
  ColoredViewState(ColoredViewState const &previousState, folly::dynamic data){};

  /*
   * Empty implementation for Android because it doesn't use this class.
   */
  folly::dynamic getDynamic() const {
    return {};
  };
  MapBuffer getMapBuffer() const {
    return MapBufferBuilder::EMPTY();
  };
#endif

 private:
  ImageSource imageSource_;
  std::shared_ptr<ImageRequest> imageRequest_;
};

} // namespace react
} // namespace facebook
```

### `ColoredViewState.cpp`
```c++
#include "ColoredViewState.h"

namespace facebook {
namespace react {

ImageSource ColoredViewState::getImageSource() const {
  return imageSource_;
}

ImageRequest const &ColoredViewState::getImageRequest() const {
  return *imageRequest_;
}

} // namespace react
} // namespace facebook
```

### `ColoredViewProps.h`
```c++
#pragma once

#include <react/renderer/components/view/ViewProps.h>
#include <react/renderer/core/PropsParserContext.h>
#include <react/renderer/imagemanager/primitives.h>

namespace facebook {
namespace react {

class ColoredViewProps final : public ViewProps {
 public:
  ColoredViewProps() = default;
  ColoredViewProps(const PropsParserContext& context, const ColoredViewProps &sourceProps, const RawProps &rawProps);

#pragma mark - Props

  std::string color{};
  ImageSource image{};
};

} // namespace react
} // namespace facebook
```

### `ColoredViewProps.cpp`

```c++
#include "ColoredViewProps.h"
#include <react/renderer/components/image/conversions.h>
#include <react/renderer/core/PropsParserContext.h>
#include <react/renderer/core/propsConversions.h>

namespace facebook {
namespace react {

ColoredViewProps::ColoredViewProps(
    const PropsParserContext &context,
    const ColoredViewProps &sourceProps,
    const RawProps &rawProps): ViewProps(context, sourceProps, rawProps),

    color(convertRawProp(context, rawProps, "color", sourceProps.color, {})),
    image(convertRawProp(context, rawProps, "image", sourceProps.image, {}))
      {}

} // namespace react
} // namespace facebook
```

### `ColoredViewEventEmitter.h`

```c++
#pragma once

#include <react/renderer/components/view/ViewEventEmitter.h>

namespace facebook {
namespace react {

class ColoredViewEventEmitter : public ViewEventEmitter {
 public:
  using ViewEventEmitter::ViewEventEmitter;
};


} // namespace react
} // namespace facebook
```

### `ColoredViewEventEmitter.cpp`

```c++
#include "ColoredViewEventEmitter.h"

namespace facebook {
namespace react {



} // namespace react
} // namespace facebook
```

### `ColoredViewShadowNode.h`

```c++
#pragma once

#include "ColoredViewProps.h"
#include "ColoredViewState.h"
#include "ColoredViewEventEmitter.h"
#include <react/renderer/components/view/ConcreteViewShadowNode.h>
#include <react/renderer/imagemanager/ImageManager.h>

namespace facebook {
namespace react {

extern const char ColoredViewComponentName[];

/*
 * `ShadowNode` for <ColoredView> component.
 */
class ColoredViewShadowNode final: public ConcreteViewShadowNode<
  ColoredViewComponentName,
  ColoredViewProps,
  ColoredViewEventEmitter,
  ColoredViewState
> {

public:
  using ConcreteViewShadowNode::ConcreteViewShadowNode;

  // Associates a shared `ImageManager` with the node.
  void setImageManager(const SharedImageManager &imageManager);

  static ColoredViewState initialStateData(
                                           ShadowNodeFragment const &fragment,
                                           ShadowNodeFamilyFragment const &familyFragment,
                                           ComponentDescriptor const &componentDescriptor) {
    auto imageSource = ImageSource{ImageSource::Type::Invalid};
    return {
      imageSource,
      {imageSource, nullptr}};
  }

#pragma mark - LayoutableShadowNode

  Size measureContent(
                      LayoutContext const &layoutContext,
                      LayoutConstraints const &layoutConstraints) const override;
  void layout(LayoutContext layoutContext) override;

private:
  void updateStateIfNeeded();

  ImageSource getImageSource() const;

  SharedImageManager imageManager_;
};

} // namespace react
} // namespace facebook
```

### `ColoredViewShadowNode.cpp`

```c++
#include "ColoredViewShadowNode.h"
#include <react/renderer/core/LayoutContext.h>

namespace facebook {
namespace react {

extern const char ColoredViewComponentName[] = "ColoredView";

void ColoredViewShadowNode::setImageManager(const SharedImageManager &imageManager) {
  ensureUnsealed();
  imageManager_ = imageManager;
}

void ColoredViewShadowNode::updateStateIfNeeded() {
  const auto &newImageSource = getImageSource();

  auto const &currentState = getStateData();

  auto imageSource = currentState.getImageSource();

  bool anyChanged = newImageSource != imageSource;

  if (!anyChanged) {
    return;
  }

  // Now we are about to mutate the Shadow Node.
  ensureUnsealed();

  // It is not possible to copy or move image requests from SliderLocalData,
  // so instead we recreate any image requests (that may already be in-flight?)
  // TODO: check if multiple requests are cached or if it's a net loss
  auto state = ColoredViewState{
      newImageSource,
      imageManager_->requestImage(newImageSource, getSurfaceId()),
    };
  setStateData(std::move(state));
}

ImageSource ColoredViewShadowNode::getImageSource() const {
  return getConcreteProps().image;
}

#pragma mark - LayoutableShadowNode

Size ColoredViewShadowNode::measureContent(
    LayoutContext const &layoutContext,
    LayoutConstraints const &layoutConstraints) const {
  return {};
}

void ColoredViewShadowNode::layout(LayoutContext layoutContext) {
  updateStateIfNeeded();
  ConcreteViewShadowNode::layout(layoutContext);
}

} // namespace react
} // namespace facebook
```

### `ColoredViewComponentDescriptor.h`

```c++

/**
 * This code was generated by [react-native-codegen](https://www.npmjs.com/package/react-native-codegen).
 *
 * Do not edit this file as changes may cause incorrect behavior and will be lost
 * once the code is regenerated.
 *
 * @generated by codegen project: GenerateComponentDescriptorH.js
 */

#pragma once

#include "ColoredViewShadowNode.h"
#include <react/renderer/core/ConcreteComponentDescriptor.h>

namespace facebook {
namespace react {

/*
 * Descriptor for <Slider> component.
 */
class ColoredViewComponentDescriptor final
    : public ConcreteComponentDescriptor<ColoredViewShadowNode> {
 public:
    ColoredViewComponentDescriptor(ComponentDescriptorParameters const &parameters)
      : ConcreteComponentDescriptor(parameters),
        imageManager_(std::make_shared<ImageManager>(contextContainer_)){}

  void adopt(ShadowNode::Unshared const &shadowNode) const override {
    ConcreteComponentDescriptor::adopt(shadowNode);

    auto coloredViewShadowNode =
        std::static_pointer_cast<ColoredViewShadowNode>(shadowNode);

    // `SliderShadowNode` uses `ImageManager` to initiate image loading and
    // communicate the loading state and results to mounting layer.
    coloredViewShadowNode->setImageManager(imageManager_);
  }

 private:
  const SharedImageManager imageManager_;
};

} // namespace react
} // namespace facebook
```

### `RCTComponentViewHelpers.h`

```C++
#import <Foundation/Foundation.h>
#import <React/RCTDefines.h>
#import <React/RCTLog.h>

NS_ASSUME_NONNULL_BEGIN

@protocol RCTColoredViewViewProtocol <NSObject>

@end

NS_ASSUME_NONNULL_END
```

----
At this point, we have all the elements that we need and we can write our component.

### `RNColoredView.h`

```obj-c
#import <React/RCTViewComponentView.h>
#import <UIKit/UIKit.h>

#ifndef NativeComponentExampleComponentView_h
#define NativeComponentExampleComponentView_h

NS_ASSUME_NONNULL_BEGIN

@interface RNColoredView : RCTViewComponentView
@end

NS_ASSUME_NONNULL_END

#endif /* NativeComponentExampleComponentView_h */
```

### `RNColoredView.cpp`

```c++
#import "RNColoredView.h"

#import <React/RCTConversions.h>
#import <React/RCTImageResponseDelegate.h>
#import <React/RCTImageResponseObserverProxy.h>

#import <react/renderer/imagemanager/ImageResponseObserverCoordinator.h>
#import "ColoredViewComponentDescriptor.h"
#import "ColoredViewProps.h"
#import "RCTComponentViewHelpers.h"

#import "RCTFabricComponentsPlugins.h"

using namespace facebook::react;

@interface RNColoredView () <RCTColoredViewViewProtocol, RCTImageResponseDelegate>

@end

@implementation RNColoredView {
    UIView * _view;
    UIImageView *_imageView;
    ImageResponseObserverCoordinator const *_imageCoordinator;
    RCTImageResponseObserverProxy _imageResponseObserverProxy;
}

+ (ComponentDescriptorProvider)componentDescriptorProvider
{
    return concreteComponentDescriptorProvider<ColoredViewComponentDescriptor>();
}

- (instancetype)initWithFrame:(CGRect)frame
{
if (self = [super initWithFrame:frame]) {
    static const auto defaultProps = std::make_shared<const ColoredViewProps>();
    _props = defaultProps;

    _view = [[UIView alloc] init];

    _imageView = [[UIImageView alloc] init];
    [_view addSubview:_imageView];

    _imageView.translatesAutoresizingMaskIntoConstraints = false;
    [NSLayoutConstraint activateConstraints:@[
      [_imageView.topAnchor constraintEqualToAnchor:_view.topAnchor constant:10],
      [_imageView.leftAnchor constraintEqualToAnchor:_view.leftAnchor constant:10],
      [_imageView.bottomAnchor constraintEqualToAnchor:_view.bottomAnchor constant:-10],
      [_imageView.rightAnchor constraintEqualToAnchor:_view.rightAnchor constant:-10],
    ]];
    _imageView.backgroundColor = UIColor.redColor;

    _imageResponseObserverProxy = RCTImageResponseObserverProxy(self);

    self.contentView = _view;
}

return self;
}

- (void)updateProps:(Props::Shared const &)props oldProps:(Props::Shared const &)oldProps
{
    const auto &oldViewProps = *std::static_pointer_cast<ColoredViewProps const>(_props);
    const auto &newViewProps = *std::static_pointer_cast<ColoredViewProps const>(props);

    if (oldViewProps.color != newViewProps.color) {
        NSString * colorToConvert = [[NSString alloc] initWithUTF8String: newViewProps.color.c_str()];
        [_view setBackgroundColor:[self hexStringToColor:colorToConvert]];
    }

    [super updateProps:props oldProps:oldProps];
}

- (void)updateState:(const facebook::react::State::Shared &)state oldState:(const facebook::react::State::Shared &)oldState
{
  auto _state = std::static_pointer_cast<ColoredViewShadowNode::ConcreteState const>(state);
  auto _oldState = std::static_pointer_cast<ColoredViewShadowNode::ConcreteState const>(oldState);

  auto data = _state->getData();

  bool havePreviousData = _oldState != nullptr;

  auto getCoordinator = [](ImageRequest const *request) -> ImageResponseObserverCoordinator const * {
    if (request) {
      return &request->getObserverCoordinator();
    } else {
      return nullptr;
    }
  };

  if (!havePreviousData || data.getImageSource() != _oldState->getData().getImageSource()) {
    self.imageCoordinator = getCoordinator(&data.getImageRequest());
  }
}

- (void)setImageCoordinator:(const ImageResponseObserverCoordinator *)coordinator
{
  if (_imageCoordinator) {
    _imageCoordinator->removeObserver(_imageResponseObserverProxy);
  }
  _imageCoordinator = coordinator;
  if (_imageCoordinator) {
    _imageCoordinator->addObserver(_imageResponseObserverProxy);
  }
}

- (void)setImage:(UIImage *)image
{
  if ([image isEqual:_imageView.image]) {
    return;
  }

  _imageView.image = image;
}

Class<RCTComponentViewProtocol> ColoredViewCls(void)
{
    return RNColoredView.class;
}

- hexStringToColor:(NSString *)stringToConvert
{
    NSString *noHashString = [stringToConvert stringByReplacingOccurrencesOfString:@"#" withString:@""];
    NSScanner *stringScanner = [NSScanner scannerWithString:noHashString];

    unsigned hex;
    if (![stringScanner scanHexInt:&hex]) return nil;
    int r = (hex >> 16) & 0xFF;
    int g = (hex >> 8) & 0xFF;
    int b = (hex) & 0xFF;

    return [UIColor colorWithRed:r / 255.0f green:g / 255.0f blue:b / 255.0f alpha:1.0f];
}

#pragma mark - RCTImageResponseDelegate

- (void)didReceiveImage:(UIImage *)image metadata:(id)metadata fromObserver:(void const *)observer
{
  if (observer == &_imageResponseObserverProxy) {
    self.image = image;
  }
}

- (void)didReceiveProgress:(float)progress fromObserver:(void const *)observer
{
}

- (void)didReceiveFailureFromObserver:(void const *)observer
{
}

@end
```

**Reply by cipolleschi** [reply]
Another alternative worth mentioning, which is probably the fastest migration path, is to use the strategy 2 (props with a custom Image loader) using the `[[RCTBridge currentBridge] moduleForName:@"ImageLoader"];` as custom `ImageLoader`.

**Reply by grahammendick** [reply]
Thanks for such a full answer. I appreciate it must've been tough to harvest all that info.

I want to try your latest suggestion of using React Native's image loader because it promises the fastest migration path. But in order to use `[RCTBridge currentBridge]` I have to import `RCTBridge+Private.h`. Is this ok?

**Reply by cipolleschi** [reply]
I would say yes, for the time being. Meanwhile, I will try to dig into the matter more: eventually, we could also expose the `currentBridge` getter in the public APIs.

**Reply by grahammendick** [reply]
Thanks, that sounds like a good way forward. I also need to import `RCTImagePrimitivesConversions.h` (to convert the image into `NSURLRequest`) and `RCTImageLoaderProtocol`. So just double checking both of those will stay publicly available.

**Reply by grahammendick** [reply]
I've unmarked it as the answer because of the `RCTBridge+Private` import. Once the `currentBridge` is exposed in the public api I'll mark this as the answer

**Comment by grahammendick** [comment, marked as answer]
I’ve spent weeks working out how to use c++/rendered state in Fabric. I don’t understand why React Native couldn’t write this up. It seems needlessly cruel of them. I initially took the 'Write You Own State' section above to be React Native’s answer on how to use c++ state. But there are mistakes and omissions. So I’m going to save other people the pain by writing up what I learnt.

The key omission from the 'Write You Own State' section above is it doesn’t mention the `interfaceOnly` property of the NativeComponent spec. When you include this then React Native doesn’t codegen everything. It still codegens the Props and EventEmitters, for example, but it doesn’t codegen the ComponentDescriptor, State or ShadowNode. 

```js
export default (codegenNativeComponent<NativeProps>(
   'NVBarButton',
   {interfaceOnly: true}
): HostComponent<NativeProps>);
```

This simplifies the 'Write You Own State' section because there are now only five classes to hand-write. These can be copied either from [React Native repo](https://github.com/facebook/react-native/tree/main/ReactCommon/react/renderer/components/slider) or from the  'Write You Own State' section above.

```
NVBarButtonComponentDescriptor.h
NVBarButtonShadowNode.cpp
NVBarButtonShadowNode.h
NVBarButtonState.cpp
NVBarButtonState.h
```

The next omission is that these files must be included as source files in the podspec for them to make it into the iOS build. So add them to a ‘cpp’ folder, for example, and then add that to the `source_files` of the podspec.
```
spec.source_files = "ios/**/*.{h,m,mm}", "cpp/**/*.{h,cpp}"
```

You can see all these changes in [my PR that adds image props](https://github.com/grahammendick/navigation/pull/607) to the `BarButton` component of my Navigation library. 

Another omission is how to get Android working with these changes. Even though c++ state isn’t needed on Android (image props are handled differently), Android won’t build without these hand-written files.

**The following changes need to be made by the user of your library in their app.**

In the Android.mk of the app

```mk
LOCAL_C_INCLUDES := $(LOCAL_PATH) $(NODE_MODULES_DIR)/navigation-react-native/cpp
LOCAL_SRC_FILES := $(wildcard $(LOCAL_PATH)/*.cpp) $(wildcard $(NODE_MODULES_DIR)/navigation-react-native/cpp/*.cpp)
LOCAL_EXPORT_C_INCLUDES := $(LOCAL_PATH)
```

And include `mapbuffer` and `imagemanager` in LOCAL_SHARED_LIBRARIES

```mk
libreact_render_mapbuffer \
libreact_render_imagemanager \
```

In MainComponentsRegistry.cpp of the app
```cpp
#include <NVBarButtonComponentDescriptor.h>
```
